watcher/watcher.go

```diff
@@ -0,0 +1,296 @@
+	if !scanNewWorkloads {
+		wh.watchPodsNoScanOnNewWorkloads(ctx)
+	} else {
+		wh.watchPodsTriggerScanOnNewWorkloads(ctx)
+	}
+}
+
+// returns a watcher watching from current resource version
+func (wh *WatchHandler) getPodWatcher() (watch.Interface, error) {
+	podsWatch, err := wh.k8sAPI.KubernetesClient.CoreV1().Pods("").Watch(context.TODO(), v1.ListOptions{
+		ResourceVersion: wh.currentResourceVersion,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return podsWatch, nil
+}
+
+func (wh *WatchHandler) restartResourceVersion(podWatch watch.Interface) error {
+	podWatch.Stop()
+	return wh.updateResourceVersion()
+}
+
+func (wh *WatchHandler) getNewImages(pod *core1.Pod) []string {
+	newImgs := make([]string, 0)
+	for _, imageID := range extractImageIDsFromPod(pod) {
+		if _, ok := wh.imagesIDsMap[imageID]; !ok {
+			wh.imagesIDsMap[imageID] = true
+			newImgs = append(newImgs, imageID)
+		}
+	}
+	return newImgs
+}
+
+// returns pod and true if event status is modified, pod is exists and is running
+func (wh *WatchHandler) getPodFromEventIfRunning(event watch.Event) (*core1.Pod, bool) {
+	if event.Type != watch.Modified {
+		return nil, false
+	}
+	var pod *core1.Pod
+	if val, ok := event.Object.(*core1.Pod); ok {
+		pod = val
+		if pod.Status.Phase != core1.PodRunning {
+			return nil, false
+		}
+	}
+
+	// check that Pod exists (when deleting a Pod we get MODIFIED events with Running status)
+	_, err := wh.k8sAPI.GetWorkload(pod.GetNamespace(), "pod", pod.GetName())
+	if err != nil {
+		return nil, false
+	}
+
+	return pod, true
+}
+
+// pod watcher that does not  trigger scan on new workloads
+func (wh *WatchHandler) watchPodsNoScanOnNewWorkloads(ctx context.Context) {
+	logger.L().Ctx(ctx).Debug("starting pod watch")
+	for {
+		podsWatch, err := wh.getPodWatcher()
+		if err != nil {
+			logger.L().Ctx(ctx).Error(fmt.Sprintf("error to getPodWatcher, err :%s", err.Error()), helpers.Error(err))
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		for {
+			event, ok := <-podsWatch.ResultChan()
+			if !ok {
+				// channel closed, restart watch
+				err = wh.restartResourceVersion(podsWatch)
+				if err != nil {
+					logger.L().Ctx(ctx).Error(fmt.Sprintf("error to restartResourceVersion, err :%s", err.Error()), helpers.Error(err))
+				}
+				break
+
+			}
+
+			if pod, ok := wh.getPodFromEventIfRunning(event); ok {
+				newImgs := wh.getNewImages(pod)
+
+				if len(newImgs) > 0 {
+					parentWlid, err := wh.getParentIDForPod(pod)
+					if err != nil {
+						logger.L().Ctx(ctx).Error(fmt.Sprintf("error to getParentIDForPod, err :%s", err.Error()), helpers.Error(err))
+						continue
+					}
+					wh.triggerSBOMCalculation(parentWlid, newImgs)
+				}
+			}
+		}
+	}
+}
+
+// pod watcher that triggers scan on new workloads
+func (wh *WatchHandler) watchPodsTriggerScanOnNewWorkloads(ctx context.Context) {
+	logger.L().Ctx(ctx).Debug("starting pod watch")
+	for {
+		podsWatch, err := wh.getPodWatcher()
```

Comment:
Let logic of  getPodWatcher and waiting on the ResultChan is duplicated - you can remove it by moving it to the calling function, and have separate logic only for the actual event handling.